src/rgw/rgw_rest_s3.cc

```diff
@@ -1809,43 +1809,65 @@ void RGWListBucket_ObjStore_S3::send_response()
     s->formatter->dump_string("EncodingType", "url");
     encode_key = true;
   }
+  
   RGWListBucket_ObjStore_S3::send_common_response();
-    if (op_ret >= 0) {
-      vector<rgw_bucket_dir_entry>::iterator iter;
-      for (iter = objs.begin(); iter != objs.end(); ++iter) {
-        rgw_obj_key key(iter->key);
-        s->formatter->open_array_section("Contents");
-        if (encode_key) {
-          string key_name;
-          url_encode(key.name, key_name);
-          s->formatter->dump_string("Key", key_name);
+  
+  if (op_ret >= 0) {
+    if (s->format == RGWFormat::JSON) {
+      s->formatter->open_array_section("Contents");
+    }
```

Comment:
`optional<Formatter::ArraySection>` might be a good way to conditionally open a section without having to match it with a conditional close. think this helps?
```c++
{
  std::optional<Formatter::ArraySection> json_array;
  if (s->format == RGWFormat::JSON) {
    json_array.emplace("Contents");
  }
  // ...
}
```

Reply:
```{
  "Name": "works3",
  "Prefix": "",
  "MaxKeys": 1000,
  "IsTruncated": "false",
  "Contents": [
    {
      "Key": "pass1",
      "LastModified": "2022-08-15T17:22:01.609Z",
      "ETag": "\"271ec5227ced4bd1f5de40e6d665b3e1\"",
      "Size": 4458,
      "StorageClass": "STANDARD",
      "Owner": {
        "ID": "testid",
        "DisplayName": "M. Tester"
      },
      "Type": "Normal"
    },
    {
      "Key": "pass2",
      "LastModified": "2022-08-16T12:53:10.667Z",
      "ETag": "\"271ec5227ced4bd1f5de40e6d665b3e1\"",
      "Size": 4458,
      "StorageClass": "STANDARD",
      "Owner": {
        "ID": "testid",
        "DisplayName": "M. Tester"
      },
      "Type": "Normal"
    },
    {
      "Key": "pass3",
      "LastModified": "2022-08-16T12:53:22.206Z",
      "ETag": "\"271ec5227ced4bd1f5de40e6d665b3e1\"",
      "Size": 4458,
      "StorageClass": "STANDARD",
      "Owner": {
        "ID": "testid",
        "DisplayName": "M. Tester"
      },
      "Type": "Normal"
    }
  ],
  "Marker": ""
}```

Reply:
what say you, @cbodley ?

Reply:
it works3!

Reply:
(the boost::optional idiom is cute, but doesn't seem simpler to me)

Reply:
it's a clever overload of raii, maybe I can get more used to that over time